app/lib/task/cloudcompute/googlecloudcompute.dart

```diff
@@ -0,0 +1,502 @@
+  return await retry(
+    fn,
+    retryIf: (e) =>
+        // Guessing the API might honor: https://google.aip.dev/194
+        // So only retry 'UNAVAILABLE' errors, which is 503 according to:
+        // https://github.com/googleapis/api-common-protos/blob/master/google/rpc/code.proto
+        (e is DetailedApiRequestError && e.status == 503) ||
+        // In addition we retry undocumented errors and malformed responses.
+        (e is ApiRequestError && e is! DetailedApiRequestError) ||
+        // If there is a timeout, we also retry.
+        e is TimeoutException ||
+        // Finally, we retry all I/O issues.
+        e is IOException,
+  );
+}
+
+/// Pattern for valid GCE instance names.
+final _validInstanceNamePattern = RegExp(r'^[a-z]([-a-z0-9]*[a-z0-9])?$');
+
+@sealed
+class _GoogleCloudCompute extends CloudCompute {
+  final ComputeApi _api;
+
+  /// GCP project this isntance is managing VMs inside.
+  final String _project;
+
+  /// GCP zones this instance is managing VMs inside.
+  final List<String> _zones;
+
+  /// GCP machine type, see:
+  /// https://cloud.google.com/compute/docs/machine-types
+  final String _machineType;
+
+  /// Value of the `pool` label for VMs managed by this instance.
+  ///
+  /// Instances created must have this `pool` label, same goes for instances
+  /// listed (luckily we can filter in labels in the API).
+  final String _poolLabel;
+
+  /// Instances where a [Future] from the [createInstance] operation is still
+  /// waiting to be resolved.
+  ///
+  /// We shall show such instances in [listInstances] until the [Future]
+  /// returned frm [createInstance] has been resolved.
+  final Set<_PendingGoogleCloudInstance> _pendingInstances = {};
+
+  _GoogleCloudCompute(
+    this._api,
+    this._project,
+    this._zones,
+    this._machineType,
+    this._poolLabel,
+  );
+
+  @override
+  List<String> get zones => List.from(_zones);
+
+  @override
+  String generateInstanceName() => 'worker-${Ulid()}';
+
+  @override
+  Future<CloudInstance> createInstance({
+    required String zone,
+    required String name,
+    required String dockerImage,
+    required List<String> arguments,
+    required String description,
+  }) async {
+    if (!_zones.contains(zone)) {
+      throw ArgumentError.value(
+        zone,
+        'zone',
+        'must be one of CloudCompute.zones',
+      );
+    }
+    if (name.isEmpty || name.length > 63) {
+      throw ArgumentError.value(
+          name, 'name', 'must have a length between 1 and 63');
+    }
+    if (!_validInstanceNamePattern.hasMatch(name)) {
+      throw ArgumentError.value(
+          name, 'name', 'must match pattern: $_validInstanceNamePattern');
+    }
+    // Max argument string size on Linux is MAX_ARG_STRLEN = 131072
+    // In addition the maximum meta-data size supported by GCE is 256KiB
+    // We need a few extra bits, so we shall enforce a max size of 100KiB.
+    if (arguments.fold<int>(0, (a, b) => a + b.length) > 100 * 1024) {
```

Comment:
Are we counting utf16 code-units or bytes here?

You might have to encode the strings to do correct counting (or ensure everything is ascii).

Reply:
Pretty sure it's bytes, so yes... It would be safer...